### PR TITLE
Whiteouts

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: utils/whiteouts/whiteouts_test.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+import (
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_servicemanager/utils/whiteouts"
```
Comment (Eabramov84) [inline utils/whiteouts/whiteouts_test.go:28]: move import "github.com/aoscloud/aos_servicemanager/utils/whiteouts" to separate section

File: utils/whiteouts/whiteouts.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		if base == whiteoutOpaqueDir {
+			if err := unix.Setxattr(dir, "trusted.overlay.opaque", []byte{'y'}, 0); err != nil {
+				return aoserrors.Wrap(err)
+			}
+
+			return nil
+		}
+
+		if strings.HasPrefix(base, whiteoutPrefix) {
+			filename := base[len(whiteoutPrefix):]
```
Comment (Eabramov84) [inline utils/whiteouts/whiteouts.go:64]: use strings.TrimPrefix

Comment (Eabramov84) [reply utils/whiteouts/whiteouts.go:64]: you could use fullPath:= filepath.Join(dir,strings.TrimPrefix(base,whiteoutPrefix))

File: utils/whiteouts/whiteouts_test.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	log.SetLevel(log.DebugLevel)
+	log.SetOutput(os.Stdout)
+}
+
+/**********************************************************************************************************************
+* Main
+**********************************************************************************************************************/
+
+func TestMain(m *testing.M) {
+	var err error
```
Comment (Eabramov84) [inline utils/whiteouts/whiteouts_test.go:58]: add empty liene after var

File: utils/whiteouts/whiteouts_test.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+	testData := []testWhiteoutsData{
+		{
+			dir:      filepath.Join(layers, "etc"),
+			file:     filepath.Join(layers, "etc", ".wh.test.txt"),
+			modePath: filepath.Join(layers, "etc", "test.txt"),
+			attrDir:  filepath.Join(layers, "etc"),
+			mode:     os.ModeCharDevice,
+			compareAttr: func(path string) error {
+				opaque := make([]byte, 128)
```
Comment (Eabramov84) [inline utils/whiteouts/whiteouts_test.go:92]: empty line

File: utils/whiteouts/whiteouts_test.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+			},
+		},
+		{
+			dir:      filepath.Join(layers, "bin"),
+			file:     filepath.Join(layers, "bin", ".wh..wh..opq"),
+			modePath: filepath.Join(layers, "bin"),
+			attrDir:  filepath.Join(layers, "bin"),
+			mode:     os.ModeDir,
+			compareAttr: func(path string) error {
+				opaque := make([]byte, 128)
```
Comment (Eabramov84) [inline utils/whiteouts/whiteouts_test.go:107]: empty line

Comment (MykolaSolyanko) [reply utils/whiteouts/whiteouts_test.go:107]: Done

File: utils/whiteouts/whiteouts_test.go (hunk 4 above)
Comment (MykolaSolyanko) [reply utils/whiteouts/whiteouts_test.go:92]: Done

File: utils/whiteouts/whiteouts_test.go (hunk 3 above)
Comment (MykolaSolyanko) [reply utils/whiteouts/whiteouts_test.go:58]: Done

File: utils/whiteouts/whiteouts.go (hunk 2 above)
Comment (MykolaSolyanko) [reply utils/whiteouts/whiteouts.go:64]: Thanks. Good advice!

File: utils/whiteouts/whiteouts_test.go (hunk 1 above)
Comment (MykolaSolyanko) [reply utils/whiteouts/whiteouts_test.go:28]: Done

**Review by Eabramov84 — APPROVED** [review]

File: utils/whiteouts/whiteouts_test.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+func TestMain(m *testing.M) {
+	var err error
+
+	if tmpDir, err = ioutil.TempDir("", "aos_"); err != nil {
+		log.Fatalf("Error create tmp folder: %v", err)
+	}
+
+	ret := m.Run()
+
+	if err := os.RemoveAll(tmpDir); err != nil {
```
Comment (al1img) [inline utils/whiteouts/whiteouts_test.go:67]: No need for new var:
```go
if err = ...
```

File: utils/whiteouts/whiteouts.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	"golang.org/x/sys/unix"
+)
+
+/***********************************************************************************************************************
+* Consts
+***********************************************************************************************************************/
+
+const (
+	whiteoutPrefix    = ".wh."
+	whiteoutOpaqueDir = whiteoutPrefix + whiteoutPrefix + ".opq"
```
Comment (al1img) [inline utils/whiteouts/whiteouts.go:35]: You mask whiteoutOpaqueDir by whiteoutPrefix. If only whiteoutPrefix has to be changed then whiteoutOpaqueDir will be changed implicitly.

File: .golangci.yml [hunk 8]
```diff
@@ -29,6 +29,7 @@ linters:
     - tparallel
     - unparam
     - wastedassign
+    - nosnakecase
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply .golangci.yml:32]: not, I should move it above. I just put it a bottom all disable linter warnings

File: layermanager/layermanager_test.go [hunk 9]
```diff
@@ -468,7 +468,11 @@ func TestRemoteDownloadLayer(t *testing.T) {
 		t.Fatalf("Can't rename directory: %v", err)
 	}
 
-	server := &http.Server{Addr: ":9000", Handler: http.FileServer(http.Dir(fileServerDir))}
+	server := &http.Server{
+		Addr:              ":9000",
+		Handler:           http.FileServer(http.Dir(fileServerDir)),
+		ReadHeaderTimeout: 10 * time.Second,
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply layermanager/layermanager_test.go:474]: yes

File: utils/whiteouts/whiteouts.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+* Public
+***********************************************************************************************************************/
+
+func OCIWhiteoutsToOverlay(path string, uid, gid int) error {
+	if err := filepath.Walk(path, func(name string, info os.FileInfo, err error) error {
+		if err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		if info.IsDir() {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply utils/whiteouts/whiteouts.go:48]: Done

File: utils/whiteouts/whiteouts_test.go (hunk 6 above)
Comment (MykolaSolyanko) [reply utils/whiteouts/whiteouts_test.go:67]: Done

File: utils/whiteouts/whiteouts.go (hunk 7 above)
Comment (MykolaSolyanko) [reply utils/whiteouts/whiteouts.go:35]: Split declaration independent

**Pull request merged**